### PR TITLE
arbennett/refactor/parallel/fixes

### PR DESCRIPTION
**Pull request by arbennett**
This pull request builds on #34 and fixes some things that were introduced in that process.  First, this fixes binary and ascii input.  Second, this adds some minimal testing, and a bunch of test data.

The first commits are a holdover from the test pull request I closed due to changes in the way the way that multiprocessing was set up.  Complete example input files are given for each of the 3 input types, which all point to data that is used in the tests directory.

### CODE REVIEW

File: metsim/metsim.py [hunk 1]
```diff
@@ -1,18 +1,49 @@
...
+import time as tm
+from getpass import getuser
+from multiprocessing import Pool
+
 import numpy as np
 import pandas as pd
 import xarray as xr
-from netCDF4 import Dataset
-from multiprocessing import Process, Manager
+# from netCDF4 import Dataset
```
Comment (jhamman) [inline metsim/metsim.py]: remove commented out import

File: metsim/metsim.py [hunk 2]
```diff
@@ -1,18 +1,49 @@
...
+import metsim.constants as cnst
+
+references = '''Thornton, P.E., and S.W. Running, 1999. An improved algorithm for estimating incident daily solar radiation from measurements of temperature, humidity, and precipitation. Agricultural and Forest Meteorology, 93:211-228.
+Kimball, J.S., S.W. Running, and R. Nemani, 1997. An improved method for estimating surface humidity from daily minimum temperature. Agricultural and Forest Meteorology, 85:87-98.
+Bohn, T. J., B. Livneh, J. W. Oyler, S. W. Running, B. Nijssen, and D. P. Lettenmaier, 2013a: Global evaluation of MT-CLIM and related algorithms for forcing of ecological and hydrological models, Agr. Forest. Meteorol., 176, 38-49, doi:10.1016/j.agrformet.2013.03.003.'''
+
+now = tm.ctime(tm.time())
+user = getuser()
+
+attrs = {'pet': {'units': 'mm d-1', 'long_name': 'potential evaporation',
```
Comment (jhamman) [inline metsim/metsim.py]: units are actually `cm d-1`

File: metsim/metsim.py [hunk 3]
```diff
@@ -48,129 +79,152 @@ def __init__(self, params: dict):
...
-        # Close everything out
-        self.queue.put("done")
-        io_process.join()
-
-    def run(self, met_data, locations, queue):
+        nprocs = MetSim.params['nprocs']
+        self.pool = Pool(processes=nprocs)
+
+        # TODO: fix the hard coded 10
+        locations = np.array_split(list(zip(self.i_idx, self.j_idx)), nprocs * 10)
```
Comment (jhamman) [inline metsim/metsim.py]: Can we implement a simple heuristic for determining chunk size here? I'm not sure exactly what that would be...

File: metsim/metsim.py [hunk 4]
```diff
@@ -179,101 +233,61 @@ def ascii_out_preprocess(self):
...
-        if not os.path.exists(out_dir):
-            os.makedirs(out_dir)
-        for l in data.keys():
-            lat, lon = l.split("_")
-            out_file = '_'.join(["forcing", str(lat), str(lon)])
-            data[l].to_csv(os.path.join(out_dir, out_file), sep='\t')
-
+        shape = self.output.dims
+        for i, j in itertools.product(range(shape['lat']), range(shape['lon'])):
+            if self.output.mask[i, j]>0:
```
Comment (jhamman) [inline metsim/metsim.py]: spaces around `>`

File: metsim/metsim.py [hunk 5]
```diff
@@ -179,101 +233,61 @@ def ascii_out_preprocess(self):
...
+                lat = self.output.lat.values[i]
+                lon = self.output.lon.values[j]
+                fname = os.path.join(self.params['out_dir'], 
+                            "forcing_{}_{}.csv".format(lat, lon))
+                self.output.isel(lat=i, lon=j).to_dataframe()[
+                        self.params['out_vars']].to_csv(fname)
```
Comment (jhamman) [inline metsim/metsim.py]: you'll save a potential copy if you select the `out_vars` before calling `to_dataframe()`.

**Comment by bartnijssen** [comment]
I'll review after #34 goes through since they are cumulative

**Pull request merged**